shell/browser/file_system_access/file_system_access_permission_context.cc

```diff
@@ -881,6 +885,26 @@ bool FileSystemAccessPermissionContext::OriginHasWriteAccess(
   return false;
 }
 
+void FileSystemAccessPermissionContext::NavigatedAwayFromOrigin(
+    const url::Origin& origin) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(sequence_checker_);
+  auto it = active_permissions_map_.find(origin);
+  // If we have no permissions for the origin, there is nothing to do.
+  if (it == active_permissions_map_.end()) {
+    return;
+  }
+
+  // Start a timer to possibly clean up permissions for this origin.
+  if (!it->second.cleanup_timer) {
+    it->second.cleanup_timer = std::make_unique<base::RetainingOneShotTimer>(
```

Comment:
When `this` is destroyed, this timer is stopped + destroyed too as a result of `active_permissions_map_` being destroyed; is that correct?

Reply:
Two more questions: 

- Does this need to be done 5 seconds later, or can we just clean up permissions as soon as we navigate away?

- It looks like if >1 origin is navigated away while the timer is pending, the first origin's timer will get reset 
 and the subsequent origins will never be `RevokeGrant()`ed.

Reply:
> When this is destroyed, this timer is stopped + destroyed too as a result of active_permissions_map_ being destroyed; is that correct?

yes, correct

> Does this need to be done 5 seconds later, or can we just clean up permissions as soon as we navigate away?

We could, but I thought it best to maintain parity with upstream here

> It looks like if >1 origin is navigated away while the timer is pending, the first origin's timer will get reset
and the subsequent origins will never be RevokeGrant()ed.

Ah - this is why Chrome uses a `base::RetainingOneShotTimer` - it's a OneShotTimer, but the callback is retained
after being executed, allowing another invocation to be scheduled with `Reset()` without specifying the callback again. `Reset` in this case schedules the callback to be executed, not removes the callback.